### ISSUE TITLE
Redirect to new_password_path if reset_password_token is expired

### ISSUE THREAD

**Issue by jeremywadsack** [plataformatec/devise#4399]
The current flow of `PasswordController#update` re-renders the `edit` view whenever there is an error resetting the password. But if the error is that the token has expired the message says "Reset password token has expired, please request a new one" — with no instructions on how to do that. It would make more sense to me to have devise redirect this to the `new_password_path` in this case. That way the user has the form in front of them to make that request.

Is there a reason not to do this? If this seems like a reasonable change, I'm happy to open a pull request with the change.

**Comment by tegon** [comment]
Seems reasonable to me, but I think we would have to introduce this in a major version.

**Comment by victorhazbun** [comment]
This is not a simple change.

**Comment by tegon** [comment]
@victorhazbun Can you explain your concerns?

**Comment by victorhazbun** [comment]
@tegon I was looking into the code that handles the redirection/rendering in the controller and I saw that there was a method which sets errors for the reset_password_token https://bit.ly/2pu8vXN my concern is the way the controller responds when the resource has errors, it does not classifies which kind of error is, it always responds with the resource https://bit.ly/2FVHuXj so in order to implement this feature there should be a class that determines what to do with the resource depending on each error (It seems like the only possible error is reset_password_token). So, I believe this is what makes it not so simple, I do not know how you guys do it in this library but I can only think about a different class that has the responsibility to handle the resource errors and depending on the error it return a different action: If errors on reset_password_token then return a redirect to x_path, and so on for each kind of error.

**Comment by jeremywadsack** [comment]
We patched by replacing [line 49](https://github.com/plataformatec/devise/blob/f39c6fd92774cb66f96f546d8d5e8281542b4e78/app/controllers/devise/passwords_controller.rb#L49) with this for Rails 4.2:

```
      if resource.errors[:reset_password_token].any? { |error| error =~ /expired/ }
        redirect_to new_password_path(resource_name), alert: resource.errors.full_messages.join("\n")
      else
        respond_with resource
      end
```

In Rails 5 we can inspect the validator key rather than inspecting the message which is a better i18n approach:

```
      if resource.errors.details[:reset_password_token].any? { |error| error[:error] == :expired }
```

**Comment by tegon** [comment]
Since this is going to be included in a major change, we can support only the Rails 5 version.
@jeremywadsack would you be willing to open a PR for this?

**Comment by jeremywadsack** [comment]
Yeah, I can do that. Might be a week or two, but I'll try to get something soon.

Next major release is dropping support for rails 4? It's still supported by Rails community and Rails 6 development just started [Jan 31](http://weblog.rubyonrails.org/2018/1/31/Rails-5-2-RC1-Active-Storage-Redis-Cache-Store-HTTP2-Early-Hints-Credentials/).

**Comment by tegon** [comment]
Maybe, we don't have a release date for the next major version and we also don't have enough features to release it yet.
So when we get to the next major version it might make sense to drop the support for Rails 4.
But if you already have the solution for it too, it might be worth it to include in the PR.

**Comment by jeremywadsack** [comment]
I'll include both.

**Comment by tegon** [comment]
Closed via #4837
